lib/p2p/Pool.ts

```diff
@@ -0,0 +1,172 @@
+import net, { Server, Socket } from 'net';
+import { EventEmitter } from 'events';
+import Peer from './Peer';
+import Hosts from './Hosts';
+import NetAddress from './NetAddress';
+import Logger from '../Logger';
+
+interface PoolConfig {
+  listen: boolean;
+  port: number;
+}
+
+/** A pool of peers for handling all network activity */
+class Pool extends EventEmitter {
+  private peers: { [ key: string ]: Peer };
+  private server: Server;
+  private logger: Logger;
+  private connected: boolean;
+  private hosts: Hosts;
+
+  constructor(private config: PoolConfig) {
+    super();
+    this.peers = {};
+    this.logger = Logger.global;
+    this.server = net.createServer();
+    this.hosts = new Hosts();
+    this.connected = false;
+
+    this._init();
+  }
+
+  public connect = async (): Promise<void> => {
+    if (this.connected) {
+      return;
+    }
+
+    await this.hosts.init();
+
+    // p2p bootstrap process here
+
+    this.listen();
+    this.connected = true;
+  }
+
+  public disconnect = async (): Promise<void> => {
+    if (!this.connected) {
+      return;
+    }
+
+    await this.unlisten();
+    this.destroyPeers();
+
+    this.connected = false;
+  }
+
+  public addOutbound = async (host: string, port: number): Promise<Peer> => {
+    const address = new NetAddress(host, port);
+    const peer = Peer.fromOutbound(address);
+
+    this.bindPeer(peer);
+
+    await peer.open();
+    this.addPeer(peer);
+
+    return peer;
+  }
+
+  public broadcastOrder = (order: any) => {
+    Object.keys(this.peers).map((key) => {
+      this.peers[key].sendOrder(order);
+    });
+  }
+
+  private _init = () => {
+    this.server.on('error', (err) => {
+      console.log('err: ' + err);
+    });
+
+    this.server.on('connection', (socket: Socket) => {
+      this.handleSocket(socket);
+    });
+
+    this.server.on('listening', () => {
+      const { address, port } = this.server.address();
+      this.logger.info(`pool server listening on ${address}:${port}`);
+    });
+  }
+
+  private addInbound = async (socket: Socket): Promise<Peer> => {
+    const peer = Peer.fromInbound(socket);
+
+    this.bindPeer(peer);
+
+    await peer.open();
+    this.addPeer(peer);
+
+    return peer;
+  }
+
+  private handleSocket = (socket: Socket) => {
+    if (!socket.remoteAddress) {
```

Comment:
Under what condition would an incoming socket not have a remoteAddress?

Reply:
I saw some projects that are using that defense. 
Also, see here: https://nodejs.org/api/net.html#net_socket_remoteaddress

I'm not sure we can get that `undefined` when the socket comes from `connection` event. But I might be wrong, and anyway, `handleSocket` theoretically can be used elsewhere.

Reply:
Yeah it sounds like it would only be `undefined` after a socket has been destroyed, and logically that makes sense, but we can leave this in if the handler may be used elsewhere. Maybe just add a comment in this `if` block that the socket is destroyed?

Reply:
Sure.